packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,208 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ISequencedDocumentMessage,
+	ISnapshotTree,
+	IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IGetPendingLocalStateProps, IRuntime } from "@fluidframework/container-definitions";
+import {
+	ITelemetryLoggerExt,
+	MonitoringContext,
+	PerformanceEvent,
+	UsageError,
+	createChildMonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { assert } from "@fluidframework/core-utils";
+import { IResolvedUrl, ISnapshot } from "@fluidframework/driver-definitions";
+import { isInstanceOfISnapshot } from "@fluidframework/driver-utils";
+import {
+	ContainerStorageAdapter,
+	ISerializableBlobContents,
+	getBlobContentsFromTree,
+} from "./containerStorageAdapter";
+import { IPendingContainerState } from "./container";
+
+export class containerStateManager {
+	private readonly savedOps: ISequencedDocumentMessage[] = [];
+	public snapshot:
+		| {
+				tree: ISnapshotTree;
+				blobs: ISerializableBlobContents;
+		  }
+		| undefined;
+	private readonly mc: MonitoringContext;
+	private readonly clientId: string | undefined;
+	private readonly offlineLoadEnabled: boolean;
+	private resolvedUrl: IResolvedUrl | undefined;
+	private runtime: IRuntime | undefined;
+	private readonly storageAdapter: ContainerStorageAdapter;
```

Comment:
i wouldn't couple this class to the ContainerStorageAdapter class, as it will make testing hard. this class probably only needs a subset, like only parts of IDocumentStorageService, so it should only take that, see AttachProcessProps

Reply:
![image](https://github.com/microsoft/FluidFramework/assets/105010181/f416a6f4-7388-4ee8-b7ef-3b05f1cd3ab1)
Those are the methods that are being used inside containerStateManager. I can send those methods from the container at construction time but I'm not sure how AttachProcessProps implementation is an example or related to this so maybe I'm not understanding the approach correctly.

Reply:
look at:
```
	/**
	 * The caller should create and or get services based on the data, and its own information.
	 * @param data - the data to create services from,
	 * the summary property being the most relevant part of the data.
	 * @returns A compatible storage service
	 */
	readonly createOrGetStorageService: (
		data: ISummaryTree | undefined,
	) => Promise<Pick<IDocumentStorageService, "createBlob" | "uploadSummaryWithContext">>;
```
the container actuall returns a ContainerStorageAdapter, but the attachment props only ask for what it needs:
```
Pick<IDocumentStorageService, "createBlob" | "uploadSummaryWithContext">
```
you should only take what you need